misc/pytorch_toolkit/noise_suppression/train.py

```diff
@@ -0,0 +1,572 @@
+
+        snr_db = rand_range(self.args.snr_min, self.args.snr_max)
+        target_db = rand_range(self.args.target_min, self.args.target_max)
+
+        def db_to_scale(db):
+            return 10 ** (db / 20)
+
+        # scale noise to get given SNR
+        scale_noise = db_to_scale(-snr_db) * rms_clean / (rms_noise + EPS)
+
+        # do not scale noise if noise only rms_clean < -80dB
+        #noise_only_w = (rms_clean < db_to_scale(-80)).float()
+        #scale_noise = 1 * noise_only_w + scale_noise * (1 - noise_only_w)
+
+        x = x_clean + x_noise * scale_noise
+        rms_x = x.pow(2).mean(-1, keepdim=True).sqrt()
+
+        target_scale = db_to_scale(target_db) / (rms_x + EPS)
+        x = x * target_scale
+        x_clean = x_clean * target_scale
+
+        # randomply clip signal
+        t = torch.rand(target_scale.shape, dtype=target_scale.dtype, device=target_scale.device)
+        clamp_mask = (t < self.args.clip_prob).float()
+        # clamp_mask = clamp_mask.float()
+
+        t = torch.rand(target_scale.shape, dtype=target_scale.dtype, device=target_scale.device)
+        max_val = x.abs()
+        max_val, _ = max_val.max(-1, keepdim=True)
+        clamp_val = max_val * (1 - clamp_mask * t * 0.5)
+        x = torch.max(x, 0 - clamp_val)
+        x = torch.min(x, clamp_val)
+        return x_clean, x_noise, x
+
+    def loss(self, epoch_fp, y_clean, Y_clean, x_clean, X_clean):
+        # calc losses
+        losses = []
+
+        def add_loss(name, val):
+            self.indicators[name].append(val.item())
+            losses.append(val)
+
+        add_loss("Lri", 100 * torch.nn.functional.l1_loss(Y_clean, X_clean))
+
+        add_loss("negsisdr", -sisdr(y_clean, x_clean).mean())
+
+        if epoch_fp < 1:
+            # [B,2,F,T] -> [2,B,F,T]
+            Y = Y_clean.transpose(0, 1)
+            X = X_clean.transpose(0, 1)
+            norm_x = (X[0] * X[0] + X[1] * X[1] + EPS).sqrt().sum()
+            norm_y = (Y[0] * Y[0] + Y[1] * Y[1] + EPS).sqrt()
+            scale_x = (norm_x + EPS).reciprocal()
+            scale_y = (norm_y + EPS).reciprocal()
+            # [B, F, T]
+            cosph = ((Y[0] * X[0] + Y[1] * X[1]) * scale_y).sum() * scale_x
+            add_loss("cosph", -100 * cosph.mean())
+
+        return sum(losses)
+
+    def train(self, model, dataset_train, epoch_start):
+        self.create_dataloader(dataset_train)
+
+        self.train_batch_size = self.args.per_gpu_train_batch_size * self.world_size
+        self.gradient_accumulation_steps = max(1, self.args.total_train_batch_size // self.train_batch_size)
+        self.total_train_batch_size = self.train_batch_size * self.gradient_accumulation_steps
+        self.steps_total = int((len(self.dataloader) // self.gradient_accumulation_steps) * self.args.num_train_epochs)
+
+        self.named_params = list(model.named_parameters())
+        self.create_optimizer_and_scheduler()
+
+        self.print_info(model, dataset_train)
```

Comment:
Please, move it to the constructor. It is strange that train function should creates something.

Reply:
done